core/lib/multivm/src/versions/vm_latest/tracers/circuits_tracer.rs

```diff
@@ -0,0 +1,181 @@
+                    }
+                    a if a == ECRECOVER_PRECOMPILE_ADDRESS => {
+                        PRECOMPILE_CALL_COMMON_FRACTION + ECRECOVER_CYCLE_FRACTION
+                    }
+                    _ => PRECOMPILE_CALL_COMMON_FRACTION,
+                }
+            }
+            Opcode::FarCall(_) => FAR_CALL_FRACTION,
+            Opcode::UMA(UMAOpcode::AuxHeapWrite | UMAOpcode::HeapWrite) => UMA_WRITE_FRACTION,
+            Opcode::UMA(
+                UMAOpcode::AuxHeapRead | UMAOpcode::HeapRead | UMAOpcode::FatPointerRead,
+            ) => UMA_READ_FRACTION,
+            Opcode::Invalid(_) => unreachable!(), // invalid opcodes are never executed
+        };
+
+        self.estimated_circuits_used += used;
+    }
+}
+
+impl<S: WriteStorage, H: HistoryMode> VmTracer<S, H> for CircuitsTracer<S> {
+    fn initialize_tracer(&mut self, state: &mut ZkSyncVmState<S, H>) {
+        self.last_decommitment_history_entry_checked = Some(
+            state
+                .decommittment_processor
+                .decommitted_code_hashes
+                .history()
+                .len(),
+        );
+
+        self.last_hot_writes_history_entry_checked = Some(state.storage.hot_writes.history().len());
+
+        self.last_hot_reads_history_entry_checked = Some(state.storage.hot_reads.history().len());
+    }
+
+    fn finish_cycle(
+        &mut self,
+        state: &mut ZkSyncVmState<S, H>,
+        _bootloader_state: &mut BootloaderState,
+    ) -> TracerExecutionStatus {
+        // Trace decommitments.
+        let last_decommitment_history_entry_checked = self
+            .last_decommitment_history_entry_checked
+            .expect("Value must be set during init");
+        let history = state
+            .decommittment_processor
+            .decommitted_code_hashes
+            .history();
+        for (_, history_event) in &history[last_decommitment_history_entry_checked..] {
+            // We assume that only insertions may happen during a single VM inspection.
+            assert!(history_event.value.is_some());
+            let bytecode_len = state
+                .decommittment_processor
+                .known_bytecodes
+                .inner()
+                .get(&history_event.key)
+                .expect("Bytecode must be known at this point")
+                .len();
+
+            // Each cycle of `CodeDecommitter` processes 64 bits.
+            let decommitter_cycles_used = bytecode_len / 4;
+            self.estimated_circuits_used +=
+                (decommitter_cycles_used as f32) * CODE_DECOMMITTER_CYCLE_FRACTION;
+        }
+        self.last_decommitment_history_entry_checked = Some(history.len());
+
+        // Process storage writes
+        let last_writes_history_entry_checked = self
+            .last_hot_writes_history_entry_checked
+            .expect("Value must be set during init");
+        let history = state.storage.hot_writes.history();
```

Comment:
Well, I think `cold_writes` is simply an incorrect name, as we store slots that were written to. To avoid confusion I offer we don't use cold/hot but `written/read_storage_keys`.

Reply:
makes sense